File: backend/src/main/java/com/covenant/monitoring/service/MonitoringResultService.java
package com.covenant.monitoring.service;

import com.covenant.monitoring.model.Covenant;
import com.covenant.monitoring.model.MonitoringResult;
import com.covenant.monitoring.model.User;
import com.covenant.monitoring.repository.MonitoringResultRepository;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;

import java.time.LocalDateTime;
import java.util.List;
import java.util.Optional;

@Service
public class MonitoringResultService {

    private final MonitoringResultRepository monitoringResultRepository;

    @Autowired
    public MonitoringResultService(MonitoringResultRepository monitoringResultRepository) {
        this.monitoringResultRepository = monitoringResultRepository;
    }

    public List<MonitoringResult> getAllMonitoringResults() {
        return monitoringResultRepository.findAll();
    }

    public Optional<MonitoringResult> getMonitoringResultById(Long id) {
        return monitoringResultRepository.findById(id);
    }

    public List<MonitoringResult> getMonitoringResultsByCovenant(Covenant covenant) {
        return monitoringResultRepository.findByCovenant(covenant);
    }

    public List<MonitoringResult> getMonitoringResultsByStatus(String status) {
        return monitoringResultRepository.findByStatus(status);
    }

    public List<MonitoringResult> getMonitoringResultsByCreatedBy(User user) {
        return monitoringResultRepository.findByCreatedBy(user);
    }

    public List<MonitoringResult> getMonitoringResultsByDateRange(LocalDateTime startDate, LocalDateTime endDate) {
        return monitoringResultRepository.findByMonitoringDateBetween(startDate, endDate);
    }

    public List<MonitoringResult> getMonitoringResultsByContractId(Long contractId) {
        return monitoringResultRepository.findByCovenant_ContractId(contractId);
    }

    public List<MonitoringResult> getLatestMonitoringResultForCovenant(Long covenantId) {
        return monitoringResultRepository.findTop1ByCovenant_IdOrderByMonitoringDateDesc(covenantId);
    }

    public MonitoringResult saveMonitoringResult(MonitoringResult monitoringResult) {
        return monitoringResultRepository.save(monitoringResult);
    }

    public void deleteMonitoringResult(Long id) {
        monitoringResultRepository.deleteById(id);
    }
}

File: frontend/covenant-monitoring/src/app/core/services/monitoring-result.service.ts
import { Injectable } from '@angular/core';
import { ApiService } from './api.service';
import { Observable } from 'rxjs';
import { MonitoringResult, Covenant } from '../models/models';

@Injectable({
  providedIn: 'root'
})
export class MonitoringResultService {
  private endpoint = 'monitoring-results';
  
  constructor(private apiService: ApiService) { }

  getAllMonitoringResults(): Observable<MonitoringResult[]> {
    return this.apiService.get<MonitoringResult[]>(this.endpoint);
  }

  getMonitoringResultById(id: number): Observable<MonitoringResult> {
    return this.apiService.getById<MonitoringResult>(this.endpoint, id);
  }

  getMonitoringResultsByCovenant(covenantId: number): Observable<MonitoringResult[]> {
    return this.apiService.get<MonitoringResult[]>(`${this.endpoint}/covenant/${covenantId}`);
  }

  getMonitoringResultsByStatus(status: string): Observable<MonitoringResult[]> {
    return this.apiService.get<MonitoringResult[]>(`${this.endpoint}/status/${status}`);
  }

  getMonitoringResultsByDateRange(startDate: string, endDate: string): Observable<MonitoringResult[]> {
    return this.apiService.get<MonitoringResult[]>(`${this.endpoint}/date-range?startDate=${startDate}&endDate=${endDate}`);
  }

  getLatestMonitoringResultForCovenant(covenantId: number): Observable<MonitoringResult[]> {
    return this.apiService.get<MonitoringResult[]>(`${this.endpoint}/latest/${covenantId}`);
  }

  createMonitoringResult(monitoringResult: MonitoringResult): Observable<MonitoringResult> {
    return this.apiService.post<MonitoringResult>(this.endpoint, monitoringResult);
  }

  updateMonitoringResult(id: number, monitoringResult: MonitoringResult): Observable<MonitoringResult> {
    return this.apiService.put<MonitoringResult>(this.endpoint, id, monitoringResult);
  }

  deleteMonitoringResult(id: number): Observable<void> {
    return this.apiService.delete(this.endpoint, id);
  }
}
